File: Priority Queue/215_KthLargestElementInArray.cs
// 215. Kth Largest Element in an Array
// Tags: priority queue (heap)
//

public class Solution 
{
    public int FindKthLargest(int[] nums, int k) 
    {
        // Edge cases.
        if(k < 1)
            return -1;
        else if (nums == null || nums.Length == 0)
            return -1;
        
        PriorityQueue<int, int> pq = new();
        
        foreach(int i in nums)
        {
            pq.Enqueue(i, i);
            
            // Keep the PQ no larger than K.
            // Since this PQ by default is a Min Heap,
            // we can ensure that the smallest values
            // will be popped off first, always.
            if(pq.Count > k)
                pq.Dequeue();
        }
        
        return pq.Peek();
    }
}

File: Priority Queue/215_KthLargestElementInArray.rs
// 215. Kth Largest Element in an Array
// Tags: priority queue (heap)
//

use std::collections::BinaryHeap;

impl Solution {
    pub fn find_kth_largest(nums: Vec<i32>, k: i32) -> i32 {
        
        // Edge cases.
        if nums.is_empty() {
            return -1;
        }
        else if k < 1 {
            return -1;
        }

        let mut k = k as usize;

        // Note: the binary_heap struct in Rust is implemented as 
        // a Max Heap.
        let mut pq = BinaryHeap::with_capacity(k);

        for el in nums.iter() {
            
            pq.push(core::cmp::Reverse(el));

            if pq.len() > k {
                pq.pop();
            }
        }
        
        *pq.peek().unwrap().0
    }
}